cloud-engineering-with-gcp/python/content/lab-03/README.md

```diff
@@ -0,0 +1,320 @@
+
+In our last lab, we deployed some static HTML files to a Google Cloud Storage bucket. In the third lab of this workshop, we're going create a Docker image to run our website on NGINX and deploy it to run on [Google Cloud Run](https://cloud.google.com/run/).
+
+## Step 1 &mdash; Create a Docker Image
+
+We'll use the [Pulumi Docker provider](https://www.pulumi.com/registry/packages/docker/) to build a Docker image we can push to Google Container Registry (GCR).
+
+Before we add anything to our project, let's make sure we configure the Docker CLI to push to GCR. We can do that by running the following command:
+
+```bash
+gcloud auth configure-docker
+```
+
+We'll also need to install the [Pulumi Command provider](https://www.pulumi.com/registry/packages/command/) so we can build our image. We can do this using `pip`. Add the following line to the bottom of your `requirements.txt`:
+
+```text
+pulumi_command>=0.0.0,<1.0.0
+```
+
+Install your dependencies using pip:
+
+```bash
+pip install -r requirements.txt
+```
+
+> If you'd like to do all of this from the command line in a single command, run the following:
+>
+> ```bash
+> echo "pulumi_command>=0.0.0,<1.0.0 >> requirements.txt && pip install -r requirements.txt`
+> ```
+
+Next, we need to import the Command provider's SDK. Add the following to the top of your `__main__.py` along with your other imports:
+
+```python
+import pulumi_command as command
+```
+
+Add the following to your `__main.py__` to build the Docker image:
+
+```python
+image_tag = pulumi.Output.concat(
+    "gcr.io/",
+    pulumi.Config("gcp").require("project"),
+    "/my-first-gcp-app",
+    ":latest"
+)
+
+docker_build = command.local.Command(
+    "docker-build",
+    create=pulumi.Output.concat(
+        "docker build --platform linux/amd64 -t ", image_tag, " wwwroot 2>&1"),
+    delete=pulumi.Output.concat("docker image rm ", image_tag, " || true"),
+)
+
+docker_push = command.local.Command(
+    "docker-push",
+    create=pulumi.Output.concat("docker push ", image_tag),
+    opts=pulumi.ResourceOptions(
+        depends_on=docker_build,
+    ),
+)
+```
+
+At this stage, your `__main.py__` file should look like this:
+
+```python
+"""A Python Pulumi program"""
+
+import os
+
+import pulumi
+import pulumi_gcp as gcp
+import pulumi_command as command
+
+bucket = gcp.storage.Bucket(
+    "website",
+    location="US"
+)
+
+acl = gcp.storage.DefaultObjectAccessControl(
+    'website',
+    bucket=bucket.name,
+    role="READER",
+    entity="allUsers"
+)
+
+for file in ["404.html", "index.html"]:
+    filepath = os.path.join("wwwroot", file)
+    gcp.storage.BucketObject(
+        file,
+        bucket=bucket.name,
+        name=file,
+        source=pulumi.FileAsset(filepath),
+        opts=pulumi.ResourceOptions(depends_on=[acl])
+    )
+
+static_site_url = pulumi.Output.concat(
+    "https://storage.googleapis.com/", bucket.name, "/index.html")
+
+pulumi.export("static_url", static_site_url)
```

Comment:
Any reason we're keeping this code? Is it used in anything else?

Reply:
Will be fixed in the next commit.